File: python/08_044.py
from collections import defaultdict

import bisect

class ConfigStore:
    def __init__(self):
        self.store = defaultdict(list)
        
    def set(self, feature, version, value):
        versions = self.store[feature]
        i = bisect.bisect_left([v for v, _ in versions], version)
        
        if i < len(versions) and versions[i][0] == version:
            versions[i] = (version, value)
        else:
            versions.insert(i, (version, value))
        
    def get(self, feature, version):
        if feature not in self.store:
            return ""
        
        versions = self.store[feature]
        i = bisect.bisect_right([v for v, _ in versions], version)
        
        if i == 0:
            return ""
        
        return versions[i-1][1]

File: java/08_044.java
import java.util.HashMap;
import java.util.Map;
import java.util.TreeMap;

/**
 * Stores versioned feature configurations
 */
class ConfigStore {
   private Map<String, TreeMap<Integer, String>> store;
   
   /**
    * Initialize empty configuration store
    */
   public ConfigStore() {
      store = new HashMap<>();
   }
   
   /**
    * Set a feature's value for a specific version
    */
   public void set(String feature, int version, String value) {
      if (!store.containsKey(feature)) {
         store.put(feature, new TreeMap<>());
      }
      store.get(feature).put(version, value);
   }
   
   /**
    * Get feature value at or before specified version
    */
   public String get(String feature, int version) {
      if (!store.containsKey(feature)) {
         return "";
      }
      
      TreeMap<Integer, String> versions = store.get(feature);
      Map.Entry<Integer, String> entry = 
            versions.floorEntry(version);
      
      if (entry == null) {
         return "";
      }
      
      return entry.getValue();
   }
}